Fix multi edit button not displayed for auxiliary storage attribute

troopa81 opened the pull request:
## Description

The multi-edit button is not displayed when attribute form deal with field coming from auxiliary storage attribute.

![aux_storage_and_multiedit](https://user-images.githubusercontent.com/14358135/62965790-08e7c900-be06-11e9-83ae-983f695f33e2.png)

This PR fixes this. 

## Checklist

<!-- Reviewing is a process done by project maintainers, mostly on a volunteer basis. We try to keep the overhead as small as possible and appreciate if you help us to do so by completing the following items. Feel free to ask in a comment if you have troubles with any of them.
-->

- [x] Commit messages are descriptive and explain the rationale for changes
- [ ] Commits which fix bugs include `Fixes #11111` at the bottom of the commit message
- [x] I have read the [QGIS Coding Standards](https://docs.qgis.org/testing/en/docs/developers_guide/codingstandards.html) and this PR complies with them
- [x] New unit tests have been added for core changes
- [x] I have run [the `scripts/prepare-commit.sh` script](https://github.com/qgis/QGIS/blob/master/.github/CONTRIBUTING.md#contributing-to-qgis) before each commit
- [x] I have evaluated whether it is appropriate for this PR to be backported, backport requests are left as label or comment

troopa81 committed:
Fix multi edit button not displayed for auxiliary storage attribute

nyalldawson commented on src/gui/qgsattributeformeditorwidget.cpp:
```diff
@@ -214,7 +215,19 @@ void QgsAttributeFormEditorWidget::updateWidgets()
   bool hasMultiEditButton = ( editPage()->layout()->indexOf( mMultiEditButton ) >= 0 );
-  bool fieldReadOnly = layer()->editFormConfig().readOnly( mWidget->fieldIdx() );
+
+  bool fieldReadOnly = false;
```
This looks almost identical to QgsAttributeForm::fieldIsEditable. Could we make a common function for this instead to avoid the extra code?

troopa81 replied:
Indeed, and [here](https://github.com/qgis/QGIS/blob/master/src/gui/attributetable/qgsattributetablemodel.cpp#L767) also.

I added a method in QgsVectorLayerUtils. Hope it's the right place

troopa81 committed:
Add a utils method fieldIsEditable

nyalldawson commented on src/gui/qgsattributeform.cpp:
```diff
@@ -2204,28 +2204,16 @@ void QgsAttributeForm::updateJoinedFields( const QgsEditorWidgetWrapper &eww )
     int srcFieldIndex;
     const QgsVectorLayerJoinInfo *info = mLayer->joinBuffer()->joinForFieldIndex( fieldIndex, mLayer->fields(), srcFieldIndex );
 
     if ( info && !info->hasUpsertOnEdit() && mMode == QgsAttributeEditorContext::AddFeatureMode )
```
Is this check needed in the new method too?

troopa81 replied:
Yes, in case info is null or not editable [here](https://github.com/troopa81/QGIS/blob/fix_multiedit_button_with_auxiliary_storage/src/core/qgsvectorlayerutils.cpp#L922).

Do you prefer several return? I just keep the method the way it was written in the first place.

nyalldawson replied:
No, I mean I think it should be in QgsVectorLayerUtils::fieldIsEditable. If upsert isn't possible then the field can only be edited if the joined feature already exists, and this will impact the multi edit mode too.

troopa81 replied:
You mean like this f3f7a77 ?

So now, the multieditbutton is visible if all selected feature are not readOnly. That means that the visibility button change upon selected feature. Is that OK?

Once we agree on the behavior, I will add some unit test.

troopa81 committed:
fieldIsEditable looks for joined feature to exist if no upsertOnEdit

troopa81 committed:
Select feature to fix the multiedit mode test

troopa81 committed:
make the layer editable to the default value constraint could be updated

troopa81 commented:
I  refactor `fieldIsEditable` in this PR, and added a `layer->isEditable()` condition to the field editable condition. It seems quite normal to me that a field is editable only if its layer is editable.

As a consequence a layer must be editable to update the default value (before this PR it was working without being editable). I fixed the test accordingly c67d011 

Can someone confirm this is the right approach?

nyalldawson commented:
> before this PR it was working without being editable

When would this actually occur?

troopa81 commented:
> When would this actually occur?

I'm not sure it could happen from the application. From the test, it happens when you evaluate default definition, your layer need to be edited now. 

From the application, I try several things and it doesn't change anything. I try default value with joined layer, one edited and the one not, and there is still the same behaviour.

nyalldawson commented:
Ok, great.

I've removed the 3.4 backport tag -- I don't think the severity of this bug offsets the risk of introducing this much new logic into the LTR release.

troopa81 commented:
Thanks @nyalldawson
